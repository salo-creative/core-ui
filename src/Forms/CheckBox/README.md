# CheckBox

This component is a standalone checkbox that can be used for situations where just one true/false value is required, for example to confirm you've read an alert or similar.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import CheckBox from '@salo/core-ui/Forms/CheckBox';
```

Implement as follows

```javascript
<Checkbox 
  label='The checkbox label'
  name='field_name'
  checked={ this.state.value }
  onChange={ e => this.setState({value: e.value}) }
/>
```