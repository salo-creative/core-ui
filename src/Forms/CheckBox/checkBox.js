import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import { Wrapper, HiddenCheckBox, StyledCheckBox, CheckBoxLabel } from './checkBox.styles';

class CheckBox extends React.Component {
  handleCheckboxChange = (e) => {
    const { checked, disabled, label, name, onChange } = this.props;
    e.preventDefault();

    if (disabled) return;
    onChange({ checked: !checked, label, name });
  }

  render() {
    const {
      checked, disabled, error, errorMessage, helperText, label, margin, name, required, radio, size
    } = this.props;
    
    return (
      <Wrapper
        margin={ margin }
      >
        <CheckBoxLabel
          onClick={ this.handleCheckboxChange }
          disabled={ disabled }
          size={ size }
        >
          <HiddenCheckBox defaultChecked={ checked } name={ name } />
          <StyledCheckBox
            className={ `styled-checkbox ${ checked ? 'checked' : '' }` }
            checked={ checked }
            borderRadius={ radio ? '50%' : '3px' }
            width={ size === 'L' ? '2.5rem' : '2rem' }
          >
            <svg viewBox='0 0 24 24'>
              { radio ? <circle cx='50%' cy='50%' r='4' fill='white' /> : <polyline points='20 6 9 17 4 12' /> }
            </svg>
          </StyledCheckBox>
          { label }{ required && <sup>*</sup> }
        </CheckBoxLabel>
        <ErrorText
          disabled={ disabled }
          error={ error }
          errorMessage={ errorMessage }
          size={ size }
        />
        <HelperText
          disabled={ disabled }
          error={ error }
          helperText={ helperText }
          size={ size }
        />
      </Wrapper>
    );
  }
}

CheckBox.defaultProps = {
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  helperText: '',
  margin: '0 0 2rem 0',
  onChange: () => null,
  radio: false,
  required: false,
  size: 'M'
};

CheckBox.propTypes = {
  checked: PropTypes.bool.isRequired,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  helperText: PropTypes.string,
  label: PropTypes.string.isRequired,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  radio: PropTypes.bool,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M'])
};

export default CheckBox;