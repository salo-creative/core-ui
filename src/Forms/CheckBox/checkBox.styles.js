import styled from 'styled-components';
import { get } from 'lodash';
import { transparentize } from 'polished';

// Container
export const Wrapper = styled.div`
  margin: ${ ({ margin }) => margin };
`;

// Label
export const CheckBoxLabel = styled.label`
  align-items: center;
  color: ${ ({ theme }) => theme.primary };
  cursor: ${ ({ disabled }) => (disabled ? 'not-allowed' : 'pointer') };
  display: flex;
  font-size: ${ ({ size }) => (size === 'L' ? '1.6rem' : '1.4rem') };
  height: ${ ({ size }) => (size === 'L' ? '2.6rem' : '2.1rem') };
  margin: ${ ({ margin }) => margin };
  opacity: ${ ({ disabled }) => (disabled ? 0.75 : 1) };
  width: fit-content;
`;

// Hidden true input field for accessibility
export const HiddenCheckBox = styled.input.attrs({ type: 'checkbox' })`
  border: 0;
  clip-path: inset(50%);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

// Styled check mark
export const StyledCheckBox = styled.div`
  background-color: ${ ({ colours, checked, theme }) => {
    if (get(colours, 'checked.background')) {
      return checked ? colours.checked.background : colours.unchecked.background;
    }
    return checked ? theme.primary : theme.paleGrey;
  } };
  border-radius: ${ ({ borderRadius }) => borderRadius };
  border: ${ ({ colours, theme, checked }) => {
    if (get(colours, 'checked.border')) {
      return `.1rem solid ${ checked ? colours.checked.border : colours.unchecked.border }`;
    }
    return (checked ? `.1rem ${ theme.primary } solid` : `.1rem ${ theme.darkGrey } solid`);
  } };
  height: ${ ({ width }) => width };
  margin-right: 1rem;
  transition: all 150ms;
  box-shadow: ${ ({ shadow }) => shadow };
  width: ${ ({ width }) => width };

  svg {
    fill: none;
    stroke: ${ ({ colours, checked }) => {
    if (get(colours, 'checked.check')) {
      return checked ? colours.checked.check : colours.unchecked.check;
    }
    return '#fff';
  } };
    stroke-width: 2px;
    visibility: ${ ({ checked }) => (checked ? 'visible' : 'hidden') }
  }
`;

export const CheckMark = styled.svg`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`;

// Grouped checkboxes
export const Group = styled.div`
  width: 100%;
  flex-direction: column;
  display: flex;
  margin: ${ ({ margin }) => margin };
`;

export const GroupWrapper = styled.div`
  padding: 1rem;
  background: ${ ({ theme }) => theme.paleGrey };
  border: 0.1rem solid;
  border-color: ${ ({ theme }) => transparentize(0.5, theme.primary) };
  border-radius: 0.4rem;
`;