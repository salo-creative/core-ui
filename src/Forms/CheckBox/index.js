import CheckBox from './checkBox';
import CheckBoxGroup from './checkBox.group';

export { CheckBoxGroup };
export default CheckBox;