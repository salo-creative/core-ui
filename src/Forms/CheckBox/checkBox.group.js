import React from 'react';
import PropTypes from 'prop-types';
import { findIndex } from 'lodash';

// COMPONENTS & STYLES
import CheckBox from './checkBox';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import Label from '../components/Label';
import { Group, GroupWrapper } from './checkBox.styles';

class CheckBoxGroup extends React.Component {
  handleFieldChange = (e) => {
    const { fields, onChange } = this.props;
    const { checked, label, name } = e;
    const i = findIndex(fields, {
      name
    });
    const updatedFields = [
      ...fields.slice(0, i),
      {
        checked, label, name
      },
      ...fields.slice(i + 1)
    ];
    onChange(updatedFields);
  }

  render() {
    const {
      className,
      disabled,
      fields,
      error,
      errorMessage,
      helperText,
      label,
      name,
      margin,
      required,
      size
    } = this.props;
    return (
      <Group
        className={ `${ className } salo-checkboxGroup` }
        margin={ margin }
      >
        { label && (
          <Label
            className='salo-checkboxGroup__label'
            error={ error }
            label={ label }
            name={ name }
            required={ required }
          />
        ) }
        <GroupWrapper className='salo-checkboxGroup__wrapper'>
          { fields.map((field, i) => {
            // Set margin equivalent to parent but hide for last item
            const checkMargin = i <= (fields.length - 2) ? margin : '0';
            return (
              <CheckBox
                className={ `salo-checkboxGroup__checkbox ${ field.checked ? 'is-checked' : '' }` }
                checked={ field.checked }
                disabled={ disabled }
                label={ field.label }
                key={ field.name }
                margin={ checkMargin }
                name={ field.name }
                onChange={ this.handleFieldChange }
                size={ size }
              />
            );
          }) }
        </GroupWrapper>
        <ErrorText
          className='salo-checkboxGroup__error-text'
          disabled={ disabled }
          error={ error }
          errorMessage={ errorMessage }
          size={ size }
        />
        <HelperText
          className='salo-checkboxGroup__helper-text'
          disabled={ disabled }
          error={ error }
          helperText={ helperText }
          size={ size }
        />
      </Group>
    );
  }
}

CheckBoxGroup.defaultProps = {
  className: null,
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  fields: [],
  helperText: '',
  label: '',
  margin: '0 0 2rem 0',
  onChange: () => null,
  required: false,
  size: 'M'
};

CheckBoxGroup.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  fields: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    checked: PropTypes.bool.isRequired,
    label: PropTypes.string.isRequired
  })),
  helperText: PropTypes.string,
  label: PropTypes.string,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M'])
};

export default CheckBoxGroup;