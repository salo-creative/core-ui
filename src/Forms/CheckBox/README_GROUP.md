# CheckBoxGroup

CheckBoxGroup takes a 'fields' prop and uses that to generate a number of CheckBox elements inside a styled container. It's intended to be used for situations in which the user can select from multiple different options at once. These checkboxes inherit fields from the array passed in from props.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import { CheckBoxGroup } from '@salo/core-ui/Forms/CheckBox';
```

Implement as follows

```javascript
<CheckBoxGroup
  name='checkbox_group_name'
  fields={ this.state.fields }
  onChange={ fields => this.setState({fields}) }
>
```

The fields prop is an array of objects with the structure

```javascript
{
  name: 'field_name',
  label: 'Checkbox label',
  checked: false
}
```