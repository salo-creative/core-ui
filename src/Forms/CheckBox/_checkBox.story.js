import React from 'react';
import {
  withKnobs, boolean, text, select, object
} from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// Tests.
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// Featured Components.
import { CheckBox, CheckBoxGroup } from '../../index';

// README
import README from './README.md';
import README_GROUP from './README_GROUP.md';

// Story logic.
export const Basic = () => {
  // Store.
  const store = new Store({
    checked: false
  });
  // Knobs.
  const label = text('Label', 'Checkbox label');
  const helperText = text('Helper text', 'Helper text');
  const error = boolean('Show error state', false);
  const required = boolean('Required field', false);
  const disabled = boolean('Show disabled state', false);
  const size = select('size', ['L', 'M'], 'M');
  return (
    <RenderWithProps store={ store }>
      <CheckBox
        checked={ store.checked }
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        label={ label }
        name='story'
        onChange={ e => store.set({
          checked: e.checked
        }) }
        required={ required }
        size={ size }
      />
    </RenderWithProps>
  );
};

Basic.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    info: {
      propTablesExclude: [RenderWithProps]
    },
    notes: README,
    jest: ['checkbox']
  }
};

export const Group = () => {
  // Store.
  const store = new Store({
    fields: [
      {
        name: 'field_1', checked: false, label: 'Field 1'
      },
      {
        name: 'field_2', checked: false, label: 'Field 2'
      },
      {
        name: 'field_3', checked: false, label: 'Field 3'
      }
    ]
  });
  // Knobs.
  const label = text('Label', 'Checkbox group label');
  const helperText = text('Helper text', 'Helper text');
  const error = boolean('Show error state', false);
  const required = boolean('Required field', false);
  const disabled = boolean('Show disabled state', false);
  const size = select('size', ['L', 'M'], 'M');
  return (
    <RenderWithProps store={ store }>
      <CheckBoxGroup
        disabled={ disabled }
        error={ error }
        fields={ store.fields }
        helperText={ helperText }
        label={ label }
        name='story'
        onChange={ fields => store.set({
          fields
        }) }
        required={ required }
        size={ size }
      />
    </RenderWithProps>
  );
};

Group.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    info: {
      propTablesExclude: [RenderWithProps]
    },
    notes: README_GROUP,
    jest: ['checkbox']
  }
};

export const CustomColours = () => {
  // Store.
  const store = new Store({
    checked: false
  });
  // Knobs.
  const label = text('Label', 'Checkbox label');
  const helperText = text('Helper text', 'Helper text');
  const colours = object('Colours', {
    checked: {
      background: 'goldenrod',
      border: 'hotpink',
      check: 'red'
    },
    unchecked: {
      background: 'white',
      border: 'rebeccapurple',
      check: 'blue'
    }
  });
  const shadow = text('shadow', '0 2px 10px 0 rgba(0,0,132,0.1)');
  const error = boolean('Show error state', false);
  const required = boolean('Required field', false);
  const disabled = boolean('Show disabled state', false);
  const size = select('size', ['L', 'M'], 'M');
  return (
    <RenderWithProps store={ store }>
      <CheckBox
        checked={ store.checked }
        colours={ colours }
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        label={ label }
        name='story'
        onChange={ e => store.set({
          checked: e.checked
        }) }
        required={ required }
        size={ size }
        shadow={ shadow }
      />
    </RenderWithProps>
  );
};

CustomColours.story = {
  name: 'Custom colours',
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    info: {
      propTablesExclude: [RenderWithProps]
    },
    notes: README,
    jest: ['checkbox']
  }
};

export default {
  title: 'Forms/CheckBox'
};