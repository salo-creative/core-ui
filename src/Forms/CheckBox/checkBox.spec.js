import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import CheckBox from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <CheckBox />
  );
  expect(container.firstChild).toMatchSnapshot();
});