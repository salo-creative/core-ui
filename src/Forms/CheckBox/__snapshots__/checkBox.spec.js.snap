// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  margin: 0 0 2rem 0;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #00ACED;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.4rem;
  height: 2.1rem;
  opacity: 1;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c2 {
  border: 0;
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c3 {
  background-color: #fafcff;
  border-radius: 3px;
  border: .1rem #6c757d solid;
  height: 2rem;
  margin-right: 1rem;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  box-shadow: none;
  width: 2rem;
}

.c3 svg {
  fill: none;
  stroke: #fff;
  stroke-width: 2px;
  visibility: hidden;
}

<div
  class="c0  salo-checkbox"
>
  <label
    class="c1 salo-checkbox__label"
  >
    <input
      class="c2"
      type="checkbox"
    />
    <div
      class="c3 salo-checkbox__input styled-checkbox "
      width="2rem"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
  </label>
</div>
`;
